File: scrapers/federal_bank_scraper.py
from scrapers.base_scraper import BaseScraper
from core.extractor import LayeredExtractor
from typing import Dict, Any

class FederalBankScraper(BaseScraper):
    """
    Scraper implementation for Federal Bank.
    """
    async def scrape(self, page: Any) -> Dict[str, Any]:
        self.logger.info("starting_federal_bank_scrape")
        
        try:
            await page.wait_for_selector("table, [role='table']", timeout=5000)
        except Exception:
            self.logger.warning("timeout_waiting_for_tables_attempting_anyway")
            
        tables = await LayeredExtractor.extract_from_page(page)
        rates = []
        for t in tables:
            parsed = LayeredExtractor.parse_extracted_table(t)
            if parsed:
                rates.extend(parsed)
                
        if not rates:
            rates = await LayeredExtractor.extract_from_unstructured_text(page)
            
        return {
            "fd_rates": rates,
            "minimum_deposit": 5000.0,
            "maximum_deposit": 30000000.0,
            "premature_withdrawal_available": True,
            "premature_withdrawal_penalty": "1.00% premature withdrawal penalty",
            "loan_against_fd_available": True,
            "tax_saver_fd_available": True,
            "tax_saver_tenure": "5 Years",
            "nomination_available": True,
            "compounding_frequency": "Quarterly",
            "last_updated_on_page": None,
            "effective_from": None,
            "effective_to": None,
            "scraper_version": "1.0.0"
        }

File: scrapers/federalBankScraper.js
import { BaseScraper } from './baseScraper.js';
import { LayeredExtractor } from '../core/extractor.js';

export class FederalBankScraper extends BaseScraper {
  async scrape(page) {
    this.logger.info("starting_federal_bank_scrape");
    let rates = [];

    try {
      await page.waitForSelector("table, [role='table']", { timeout: 5000 });
    } catch (e) {
      this.logger.warn("timeout_waiting_for_tables_attempting_anyway");
    }

    // The Federal page renders several rate tables (standard retail, a long-term
    // continuation, and a special <₹3 Cr scheme). Use only the primary retail
    // table to avoid pulling in overlapping/duplicate rows from the others.
    const tables = await LayeredExtractor.extractFromPage(page);
    rates = LayeredExtractor.extractPrimaryRateRows(tables);

    if (rates.length === 0) {
      rates = await LayeredExtractor.extractFromUnstructuredText(page);
    }

    return {
      fd_rates: rates,
      minimum_deposit: 5000.0,
      maximum_deposit: 30000000.0,
      premature_withdrawal_available: true,
      premature_withdrawal_penalty: "1.00% premature withdrawal penalty",
      loan_against_fd_available: true,
      tax_saver_fd_available: true,
      tax_saver_tenure: "5 Years",
      nomination_available: true,
      compounding_frequency: "Quarterly",
      last_updated_on_page: null,
      effective_from: null,
      effective_to: null,
      scraper_version: "1.0.0"
    };
  }
}
